Advanced Copy Options - Symlinks and more!

Ryex opened the pull request:
Lets let users copy instances with some more advanced settings, potentially preserving disk-space.

Symbolic links and Hard links are the standard,  but I was even able to do CoW on ZFS, BTRFS, XFS, APFS on MacOS, and even BTRFS (via winbrtfs driver) and ReFS on windows!

fix: #761

related: #422, #38

- [x] Windows
  - [x] Filesystem::create_link (hard links) (no perms needed!)
  - [x] Filesystem::create_link (symbolic) (need elevated process)
  - [x] Filesystem::clone  ([winbtrfs](https://github.com/maharmstone/btrfs) and ReFS only, clone function uses  `FSCTL_DUPLICATE_EXTENTS_TO_FILE` and `DeviceIoControl` )
- [x] Linux
  - [x] Filesystem::create_link (hard links)
  - [x] Filesystem::create_link (symbolic)
  - [x] Filesystem::clone  (COW  via [ioctl_ficlone](https://man7.org/linux/man-pages/man2/ioctl_ficlone.2.html))
- [x] Macos
  - [x] Filesystem::create_link (hard links)
  - [x] Filesystem::create_link (symbolic)
  - [x] Filesystem::clone  (COW  via [clonefile](https://opensource.apple.com/source/xnu/xnu-3789.1.32/bsd/sys/clonefile.h.auto.html))
- [x] Export process, knows to follow symlinks.
- [x] safety feature so users know when they are modifying linked content.

Ryex committed:
we want to make links!

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
working outside windows

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
qt5 compatability

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix CMakeLits.txt for non MSVC windows builds

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
feat: successful process elevation and comunication!

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
refactor: make complete list of links to make and send that.

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
feat(symlinks&hardlinks): linkup copy dialog

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
feat(symlinks): make recursive links explicit

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix: follow symlinks when exporting

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
feat: reflink / Clone support!

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
feat(reflink): hook up relink / clone on the copy dialog

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex commented:
OK. basic functionality is in place. Now there just needs to be a discussion of the safety feature that need to be in play.

![image](https://user-images.githubusercontent.com/508861/217765953-12591df5-8ab8-442c-8ad2-5725ac059e22.png)

please review!

Ryex commented:
OK. basic functionality is in place. Now there just needs to be a discussion of the safety feature that need to be in play.

![image](https://user-images.githubusercontent.com/508861/217765953-12591df5-8ab8-442c-8ad2-5725ac059e22.png)

please review!

DioEgizio commented on launcher/DesktopServices.cpp:
```diff
@@ -37,7 +37,7 @@
 #include <QProcess>
 #include <QDebug>
-#include "Application.h"
+//#include "Application.h"
```
?

Ryex replied:
sorry, this header isn't needed here was was causing link problems with the filelink exe. I commented it out to be sure nothing else broke before removing it but forgot about it.

DioEgizio commented on launcher/CMakeLists.txt:
```diff
@@ -1093,6 +1105,52 @@ install(TARGETS ${Launcher_Name}
+    # may be unnessacery with manifest
+    if(CMAKE_GENERATOR MATCHES "Visual Studio")
+        SET_TARGET_PROPERTIES("${Launcher_Name}_filelink" PROPERTIES LINK_FLAGS "/level='requireAdministrator' /uiAccess='false' /SUBSYSTEM:CONSOLE")
+    # else()  # link arg /MANIFESTUAC only works with MSVC
```
how does it work on mingw then?

Ryex replied:
Technically I'm not even sure using this linker arg adds anything meaningful. asking the exe to request admin privileges on run doesn't work in our use context. This support *should* be added by the `.manifest` anyway so the use of the linker arg is redundant.

DioEgizio replied:
If it's already on the manifest can't we just yeet that?

Ryex replied:
Yeah, I'll remove it in the next commits I push.

DioEgizio commented on launcher/ui/dialogs/CopyInstanceDialog.ui:
```diff
@@ -136,70 +136,181 @@
+     <item>
+      <widget class="QGroupBox" name="horizontalGroupBox">
+       <property name="title">
+        <string>Clone / Reflink (Copy On Write) Options</string>
```
```suggestion
        <string>CoW (Copy-on-Write) Options</string>
```

DioEgizio commented on launcher/ui/dialogs/CopyInstanceDialog.ui:
```diff
@@ -136,70 +136,181 @@
+           <bool>false</bool>
+          </property>
+          <property name="text">
+           <string>Use Clone / Reflink</string>
```
```suggestion
           <string>Clone instead of copying</string>
```

DioEgizio commented on launcher/ui/dialogs/CopyInstanceDialog.ui:
```diff
@@ -136,70 +136,181 @@
+           </sizepolicy>
+          </property>
+          <property name="text">
+           <string>Clone / Reflink not supported on this filesystem</string>
```
```suggestion
           <string>Your filesystem and/or OS doesn't support reflinks</string>
```

DioEgizio commented on launcher/ui/dialogs/CopyInstanceDialog.cpp:
```diff
@@ -85,6 +86,22 @@ CopyInstanceDialog::CopyInstanceDialog(InstancePtr original, QWidget *parent)
+    m_cloneSupported = FS::canCloneOnFS(detectedOS);
+
+    if (m_cloneSupported) {
+        ui->cloneSupportedLabel->setText(tr("Clone / Reflink is supported on (%1)").arg(FS::getFilesystemTypeName(detectedOS)));
```
```suggestion
        ui->cloneSupportedLabel->setText(tr("Reflinks are supported on %1").arg(FS::getFilesystemTypeName(detectedOS)));
```

DioEgizio commented on launcher/ui/dialogs/CopyInstanceDialog.cpp:
```diff
@@ -85,6 +86,22 @@ CopyInstanceDialog::CopyInstanceDialog(InstancePtr original, QWidget *parent)
+    if (m_cloneSupported) {
+        ui->cloneSupportedLabel->setText(tr("Clone / Reflink is supported on (%1)").arg(FS::getFilesystemTypeName(detectedOS)));
+    } else {
+        ui->cloneSupportedLabel->setText(tr("Clone / Reflink not supported on (%1)").arg(FS::getFilesystemTypeName(detectedOS)));
```
```suggestion
        ui->cloneSupportedLabel->setText(tr("Reflinks aren't supported on %1").arg(FS::getFilesystemTypeName(detectedOS)));
```

Ryex committed:
fix: cleanup UI, detect FAT and turn off links

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix: intelegent recursive links & symlink follow on export

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex commented:
I updated the logic on links
Now there is no danger of an instance overwriting the original config when "copied" via symbolic or hard links
if "link recursively" is turned off `instance-folder/*` is linked to `new-instance-folder/*` (top level items are linked except for `instance.cfg`)
if "link recursively" is turned on `instance-folder/**/*` is linked to `new-instance-folder/**/*` (all items are linked except for `instance.cfg`)

UI logic is cleaned up:
 soft and hard links are mutually exclusive check-boxes now. Both are exclusive to clone.

Ryex committed:
fix: Qt < 5.14.0 compat

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix: better test compareison (also qt5 compat)

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix: windows test compat
fix: compiler warning on int qint32 compare

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
feat: support reflink on windows via winbtrfs!
https://github.com/maharmstone/btrfs

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex commented:
windows btrfs reflink support by calling `ReflinkCopyW` from `shellbtrfs.dll` provided by winbtrfs!

NOTE: This is slight shenanigans because `ReflinkCopyW` is a bit broken atm
https://github.com/maharmstone/btrfs/issues/556

but the code should be transparently backwards compatible if it get fixed.

Ryex committed:
fix: macos compat after refactor of `clonefile`

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
feat(reflink):  ioctl_clone for winbtrfs & ReFS

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix: add missing mingw defs

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex commented:
I'v now implimented windows Reflink on both winbtrfs and ReFS via [DeviceIoControl](https://learn.microsoft.com/en-us/windows/win32/api/ioapiset/nf-ioapiset-deviceiocontrol) calls.

The old code remains in place for now until we can be confident there are no issues.

Ryex commented:
Latest change adds the UAC shield to the symlink checkbox on windows. selecting it adds the shield icon to the ok button too.

![image](https://user-images.githubusercontent.com/508861/218295333-344a7b58-5a79-4f08-98e1-be904bcbcf27.png)

Ryex committed:
feat: Add UAC icon when symlinking on windows.

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
feat: warnings when instance resources are linked

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex commented:
and... done.

now there are warnings when resources are symbolically linked or hard linked.

![image](https://user-images.githubusercontent.com/508861/218303938-d646958c-d1f0-420d-9e3f-035f7bfe11a2.png)

DioEgizio commented on launcher/minecraft/WorldList.cpp:
```diff
@@ -207,6 +211,14 @@ QVariant WorldList::data(const QModelIndex &index, int role) const
 
+        case InfoColumn:
+            if (world.isSymLinkUnder(instDirPath())) {
+                return tr("This world is symbolicly linked from elsewhere.");
```
You mean symbolically linked ?

DioEgizio replied:
```suggestion
                return tr("This world is symbolically linked from elsewhere.");
```

DioEgizio commented on launcher/minecraft/WorldList.cpp:
```diff
@@ -222,7 +234,16 @@ QVariant WorldList::data(const QModelIndex &index, int role) const
+    {   
+        if (column == InfoColumn) {
+            if (world.isSymLinkUnder(instDirPath())) {
+                return tr("Warning: This world is symbolicly linked from elsewhere. Editing it will also change the origonal") +
```
You mean symbolically linked and editing will also change the original?

DioEgizio replied:
```suggestion
                return tr("Warning: This world is symbolically linked from elsewhere. Editing it will also change the original.") +
```

DioEgizio commented on launcher/minecraft/WorldList.cpp:
```diff
@@ -222,7 +234,16 @@ QVariant WorldList::data(const QModelIndex &index, int role) const
+                       tr("\nCanonical Path: %1").arg(world.canonicalFilePath());
+            }
+            if (world.isMoreThanOneHardLink()) {
+                return tr("Warning: This world is hard linked elsewhere. Editing it will also change the origonal");
```
original?

DioEgizio replied:
```suggestion
                return tr("Warning: This world is hard linked elsewhere. Editing it will also change the original.");
```

DioEgizio commented on launcher/minecraft/mod/ModFolderModel.cpp:
```diff
@@ -97,8 +101,24 @@ QVariant ModFolderModel::data(const QModelIndex &index, int role) const
+        if (column == NAME_COLUMN) {
+            if (at(row)->isSymLinkUnder(instDirPath())) {
+                return m_resources[row]->internal_id() +
+                    tr("\nWarning: This resource is symbolicly linked from elsewhere. Editing it will also change the origonal") +
```
Look above

DioEgizio replied:
```suggestion
                    tr("\nWarning: This resource is symbolically inked from elsewhere. Editing it will also change the original.") +
```

DioEgizio commented on launcher/minecraft/mod/ModFolderModel.cpp:
```diff
@@ -97,8 +101,24 @@ QVariant ModFolderModel::data(const QModelIndex &index, int role) const
+            }
+            if (at(row)->isMoreThanOneHardLink()) {
+                return m_resources[row]->internal_id() +
+                    tr("\nWarning: This resource is hard linked elsewhere. Editing it will also change the origonal");
```
Look above

DioEgizio replied:
```suggestion
                    tr("\nWarning: This resource is hard linked elsewhere. Editing it will also change the original.");
```

DioEgizio commented on launcher/minecraft/mod/ResourceFolderModel.cpp:
```diff
@@ -415,7 +419,25 @@ QVariant ResourceFolderModel::data(const QModelIndex& index, int role) const
+            if (column == NAME_COLUMN) {
+                if (at(row).isSymLinkUnder(instDirPath())) {
+                    return m_resources[row]->internal_id() +
+                        tr("\nWarning: This resource is symbolicly linked from elsewhere. Editing it will also change the origonal") +
```
.

DioEgizio replied:
```suggestion
                        tr("\nWarning: This resource is symbolically linked from elsewhere. Editing it will also change the original.") +
```

DioEgizio commented on launcher/minecraft/mod/ResourcePackFolderModel.cpp:
```diff
@@ -78,12 +82,28 @@ QVariant ResourcePackFolderModel::data(const QModelIndex& index, int role) const
+            if (column == NAME_COLUMN) {
+                if (at(row)->isSymLinkUnder(instDirPath())) {
+                    return m_resources[row]->internal_id() +
+                        tr("\nWarning: This resource is symbolicly linked from elsewhere. Editing it will also change the origonal") +
```
.

DioEgizio replied:
```suggestion
                        tr("\nWarning: This resource is symbolically linked from elsewhere. Editing it will also change the original.") +
```

DioEgizio commented on launcher/minecraft/mod/ResourcePackFolderModel.cpp:
```diff
@@ -78,12 +82,28 @@ QVariant ResourcePackFolderModel::data(const QModelIndex& index, int role) const
+                }
+                if (at(row)->isMoreThanOneHardLink()) {
+                    return m_resources[row]->internal_id() +
+                        tr("\nWarning: This resource is hard linked elsewhere. Editing it will also change the origonal");
```
.

DioEgizio replied:
```suggestion
                        tr("\nWarning: This resource is hard linked elsewhere. Editing it will also change the original.");
```

Ryex committed:
fix: typos

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

DioEgizio commented on launcher/ui/dialogs/CopyInstanceDialog.ui:
```diff
@@ -136,70 +136,223 @@
+          <item row="2" column="0">
+           <widget class="QCheckBox" name="symbolicLinksCheckbox">
+            <property name="text">
+             <string>Use symbloic links</string>
```
```suggestion
             <string>Use symbolic links</string>
```

Symbloic links :upside_down_face:

Ryex committed:
refactor: cleanupFilesystem.cpp
* remove now redundant reflink/clone code for windows
* remove unnessacery debug code that could slow things down

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix: windows compile broke
move winapi defs into #ifndef blocks don't check explicitly for __mingw__ define
function name win_ioctl_clone didn't get updated in teh last commit

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix: mingw still missing typedefs

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

getchoo commented:
works as expected on ntfs, but i think i came across a bit of an issue: currently, if you try to delete the instance that a copy is symbolically linked to, the copied instance is now completely unusable due to the links being unresolved. imo, a good solution here would be to delete the symlinks and copy over the files from the original instance being deleted - or at least add a warning to the deletion dialog ;)

this is awesome though, i love it!

edit: forgor to specify symbolic links

Ryex commented:
> works as expected on ntfs, but i think i came across a bit of an issue: currently, if you try to delete the instance that a copy is symbolically linked to, the copied instance is now completely unusable due to the links being unresolved. imo, a good solution here would be to delete the symlinks and copy over the files from the original instance being deleted - or at least add a warning to the deletion dialog ;)
> 
> this is awesome though, i love it!
> 
> edit: forgor to specify symbolic links

I'm not really sure what to do in this situation. It's nigh impossible to detect when a symlink exists without knowing exactly where to look for it.  The only  way to warn users when they are deleting an instance another instance links files form would be to keep track of all the symlinks an instance has and do a search through all other instances when doing a delete... I'm not sure that would be smart or quick.

getchoo commented:
> > works as expected on ntfs, but i think i came across a bit of an issue: currently, if you try to delete the instance that a copy is symbolically linked to, the copied instance is now completely unusable due to the links being unresolved. imo, a good solution here would be to delete the symlinks and copy over the files from the original instance being deleted - or at least add a warning to the deletion dialog ;)
> 
> > 
> 
> > this is awesome though, i love it!
> 
> > 
> 
> > edit: forgor to specify symbolic links
> 
> 
> 
> I'm not really sure what to do in this situation. It's nigh impossible to detect when a symlink exists without knowing exactly where to look for it.  The only  way to warn users when they are deleting an instance another instance links files form would be to keep track of all the symlinks an instance has and do a search through all other instances when doing a delete... I'm not sure that would be smart or quick.

i can understand speed when recursing through all instances, but keeping track of links not being smart? i would have to disagree with that
at bare minimum, warning users that other instances could be effected by deleting another would be an infinitely better solution than having an instance break without warning - and though you could say symbolic links make that self explanatory, a majority of people may not understand how they work or the consequences of deleting the original file.

i think that a great solution here would be to add a key to instance.cfg (if possible) that contains the source of an instance using symbolic links. this would allow the ability to easily recurse through all instances in a performance friendly way, while limiting the more time consuming task of finding symlinks. afterwards, a prompt could be shown explaining something along the lines of "the files of this instance are used by <x>, would you like to delete it as well or copy the files over?"

just an idea though, i'm not sure how hard that would be to implement - but again, at bare minimum, i think there needs to be a warning here or else there's a good chance people using this feature are going to come across this and be very confused about what comes up in the error log

TayouVR commented:
maybe give the original a list in the instance.cfg with all symlinked instances (their IDs), and then when deleting there could be a warning popup with 
"The Instance(s) %1 might reference files in this instance. Deleting it could break the other instance(s), are you sure?"

Ryex committed:
fix: ensure filelink.exe is included in setup.exe

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
feat: track instance copies that use links
confirm deleations when other instances link to it

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex commented:
@getchoo  @TayouVR 
I've added a somewhat naive link tracking system. each instance when it is made via copy saves a `linkedInstancesList` setting with is a json list containing the ID of the original instance.
When an instance is asked to be deleted there is now an additional second message box that lists the instances that say they are linked to the deletion candidate, warns that these instances may break, and asks for confirmation.

@DioEgizio I made sure the `_filelink.exe` was included in the setup exe, I know you mentioned that somewhere.

Ryex committed:
feat: connect new help button help-pages/instance-copy

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex commented:
Baring any additional review changes This should be truly done now.

Ryex committed:
fix: use noexcept overload of std::filesystem::hard_link_count

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

flowln commented on launcher/ui/dialogs/CopyInstanceDialog.ui:
```diff
@@ -136,70 +136,223 @@
+             <bool>true</bool>
+            </property>
+            <property name="toolTip">
+             <string>Use hard links instead of symbolic links.</string>
```
```suggestion
             <string>Use hard links instead of copying files.</string>
```

flowln commented on launcher/ui/dialogs/CopyInstanceDialog.ui:
```diff
@@ -136,70 +136,223 @@
+            </property>
+            <property name="text">
+             <string>Link files recursively</string>
+            </property>
+           </widget>
```
```suggestion
            </property>
            <property name="toolTip">
             <string>Link each resource individually instead of linking whole folders at once</string>
            </property>
           </widget>
```

flowln commented on launcher/ui/dialogs/CopyInstanceDialog.ui:
```diff
@@ -136,70 +136,223 @@
+           </widget>
+          </item>
+          <item row="2" column="0">
+           <widget class="QCheckBox" name="symbolicLinksCheckbox">
```
this and the hard link checkboxes could be made into radio buttons, since they are exclusive. They could also be toggled with the group box you added

flowln commented on launcher/ui/dialogs/CopyInstanceDialog.ui:
```diff
@@ -136,70 +136,223 @@
-     </item>
-     <item row="5" column="0">
-      <widget class="QCheckBox" name="copyGameOptionsCheckbox">
+    <widget class="QGroupBox" name="copyOptionsGroup">
```
you forgor the `Select All` button outside

also, would it make sense to make these groupboxes toolboxes instead? This way the dialog wouldn't need to get too much bigger the more options get added

Ryex replied:
I've tried the radio button with toggled group option. it... doesn't look good on windows and the display is inconsistent across platforms.

as for the tool boxes approach... I'm not a fan? personally I think it makes things confusing to navigate.

flowln replied:
it's fine not making it toolboxes, though at some point we will probably have to add scrolling or something to the dialog if options keep getting added.

for the radio button with toggled group option, it's a shame that's the case, i find it really counter-intuitive to have this sort of behavior in checkboxes, but if it looks bad, i guess it's fine

Ryex replied:
Yeah, on windows the group checkbox is just *there* with no regards for alignment etc.

flowln commented on launcher/FileSystem.cpp:
```diff
@@ -572,4 +980,509 @@ bool overrideFolder(QString overwritten_path, QString override_path)
+
+    QStorageInfo storage_info(NearestExistentAncestor(path));
+
+    info.fsTypeName = QString::fromStdString(storage_info.fileSystemType().toStdString());
```
```suggestion
    info.fsTypeName = storage_info.fileSystemType();
```

Ryex replied:
I'm sure there was a reason I was using a std::string at some point but clearly it's be refactored out now.

flowln commented on launcher/FileSystem.cpp:
```diff
@@ -35,6 +36,12 @@
+#include <qdebug.h>
+#include <qfileinfo.h>
+#include <qnamespace.h>
+#include <qstorageinfo.h>
```
automatic header insertion moment

flowln commented on launcher/FileSystem.cpp:
```diff
@@ -215,6 +307,275 @@ bool copy::operator()(const QString& offset, bool dryRun)
+    return true;
+}
+
+void create_link::runPrivlaged(const QString& offset)
```
```suggestion
void create_link::runPrivileged(const QString& offset)
```

probably needs a good `:%s/rivlaged/rivileged/g` on the file and header :P

flowln commented on launcher/StringUtils.cpp:
```diff
@@ -74,3 +76,16 @@ int StringUtils::naturalCompare(const QString& s1, const QString& s2, Qt::CaseSe
     return QString::compare(s1, s2, cs);
 }
+
+QString StringUtils::getRandomAlphaNumeric(const int length)
```
can't this just be a call to `QUuid::createUuid().toString()`? I doubt the length matters much in this case, so long as it provides uniqueness.

Ryex replied:
This would add "{}" to the beginning and end which would not be ideal for passing around as windows filename characters. those could be stripped like I did elsewhere. but a UUID seems *really* overkill for getting a temporary unique named pipe.

flowln replied:
For the braces, you can control the behavior of `toString` with [`this handy enum`](https://doc.qt.io/qt-6/quuid.html#StringFormat-enum).

And well, the purpose of UUID is to have a unique thingy for you to use, so why would it be overkill to use it for what it's intended? xD

Ryex replied:
I suppose it doesn't matter *that* much, and it's a much stronger guarantee of being unique (not that it needs to *be* that strong)
I wish I had read the docs more thoroughly, that enum is nice.

flowln reviewed (commented):
preliminary review, i may be able to finish it off in the weekend :slightly_smiling_face:

Ryex committed:
fix: clean up initial review comments (flowin)

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
refactor: use UUID toString mode

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Scrumplex commented:
As we introduced code signing in #865, you will need to add the filelink binary to the SignTool command here: https://github.com/PrismLauncher/PrismLauncher/blob/8ce8055aeb3e610088ff5a08f509862c1feafd8f/.github/workflows/build.yml#L399

Scrumplex commented on launcher/ui/dialogs/CopyInstanceDialog.cpp:
```diff
@@ -220,3 +289,39 @@ void CopyInstanceDialog::on_copyScreenshotsCheckbox_stateChanged(int state)
+    m_selectedOptions.enableUseClone(m_cloneSupported && (state == Qt::Checked));
+    updateUseCloneCheckbox();
+    updateLinkOptions();
+}
```
newline

Scrumplex commented on launcher/BaseInstance.cpp:
```diff
@@ -64,6 +66,8 @@ BaseInstance::BaseInstance(SettingsObjectPtr globalSettings, SettingsObjectPtr s
     m_settings->registerSetting("totalTimePlayed", 0);
     m_settings->registerSetting("lastTimePlayed", 0);
 
+    m_settings->registerSetting("linkedInstancesList", "[]");
```
```suggestion
    m_settings->registerSetting("linkedInstances", "[]");
```

Scrumplex commented on launcher/BaseInstance.cpp:
```diff
@@ -182,6 +186,38 @@ bool BaseInstance::shouldStopOnConsoleOverflow() const
 
+QStringList BaseInstance::getLinkedInstances() const
+{
+    return m_settings->getList<QString>("linkedInstancesList");
```
```suggestion
    return m_settings->getList<QString>("linkedInstances");
```

Scrumplex commented on launcher/BaseInstance.cpp:
```diff
@@ -182,6 +186,38 @@ bool BaseInstance::shouldStopOnConsoleOverflow() const
+
+void BaseInstance::setLinkedInstances(const QStringList& list)
+{
+    auto linkedInstancesList = m_settings->getList<QString>("linkedInstancesList");
+    m_settings->setList("linkedInstancesList", list);
```
```suggestion
    auto linkedInstancesList = m_settings->getList<QString>("linkedInstances");
    m_settings->setList("linkedInstances", list);
```

Scrumplex commented on launcher/BaseInstance.cpp:
```diff
@@ -182,6 +186,38 @@ bool BaseInstance::shouldStopOnConsoleOverflow() const
+
+void BaseInstance::addLinkedInstanceId(const QString& id)
+{
+    auto linkedInstancesList = m_settings->getList<QString>("linkedInstancesList");
```
```suggestion
    auto linkedInstancesList = m_settings->getList<QString>("linkedInstances");
```

Scrumplex commented on launcher/BaseInstance.cpp:
```diff
@@ -182,6 +186,38 @@ bool BaseInstance::shouldStopOnConsoleOverflow() const
+
+bool BaseInstance::removeLinkedInstanceId(const QString& id)
+{
+    auto linkedInstancesList = m_settings->getList<QString>("linkedInstancesList");
```
```suggestion
    auto linkedInstancesList = m_settings->getList<QString>("linkedInstances");
```

Scrumplex commented on launcher/BaseInstance.cpp:
```diff
@@ -182,6 +186,38 @@ bool BaseInstance::shouldStopOnConsoleOverflow() const
+
+bool BaseInstance::isLinkedToInstanceId(const QString& id) const
+{
+    auto linkedInstancesList = m_settings->getList<QString>("linkedInstancesList");
```
```suggestion
    auto linkedInstancesList = m_settings->getList<QString>("linkedInstances");
```

Scrumplex commented on launcher/InstanceCopyTask.cpp:
```diff
@@ -4,13 +4,30 @@
     m_origInstance = origInstance;
     m_keepPlaytime = prefs.isKeepPlaytimeEnabled();
 
+    
+
+    
```
why so many newlines

Scrumplex commented on launcher/filelink/main.cpp:
```diff
@@ -0,0 +1,31 @@
+    FileLinkApp ldh(argc, argv);
+ 
+    return ldh.exec();
+}
```
newline

Scrumplex commented on launcher/filelink/filelink.exe.manifest:
```diff
@@ -0,0 +1,28 @@
+       </requestedPrivileges>
+    </security>
+  </trustInfo>
+</assembly>
```
newline

Scrumplex commented on launcher/minecraft/mod/ModFolderModel.cpp:
```diff
@@ -97,8 +101,24 @@ QVariant ModFolderModel::data(const QModelIndex &index, int role) const
     case Qt::ToolTipRole:
+        if (column == NAME_COLUMN) {
+            if (at(row)->isSymLinkUnder(instDirPath())) {
+                return m_resources[row]->internal_id() +
+                    tr("\nWarning: This resource is symbolically linked from elsewhere. Editing it will also change the original") +
+                    tr("\nCanonical Path: %1").arg(at(row)->fileinfo().canonicalFilePath());
```
wrap all of this into a single tr()

Scrumplex commented on launcher/minecraft/mod/ModFolderModel.cpp:
```diff
@@ -97,8 +101,24 @@ QVariant ModFolderModel::data(const QModelIndex &index, int role) const
+                    tr("\nCanonical Path: %1").arg(at(row)->fileinfo().canonicalFilePath());
+            }
+            if (at(row)->isMoreThanOneHardLink()) {
+                return m_resources[row]->internal_id() +
```
same here

Scrumplex commented on launcher/minecraft/WorldList.cpp:
```diff
@@ -222,7 +234,16 @@ QVariant WorldList::data(const QModelIndex &index, int role) const
+    {   
+        if (column == InfoColumn) {
+            if (world.isSymLinkUnder(instDirPath())) {
+                return tr("Warning: This world is symbolically linked from elsewhere. Editing it will also change the original") +
+                       tr("\nCanonical Path: %1").arg(world.canonicalFilePath());
```
same here

Scrumplex commented on launcher/minecraft/mod/ResourceFolderModel.h:
```diff
@@ -125,6 +125,8 @@ class ResourceFolderModel : public QAbstractListModel {
         [[nodiscard]] bool lessThan(const QModelIndex& source_left, const QModelIndex& source_right) const override;
     };
 
+    QString instDirPath() const;
```
maybe just make this a parameter for the constructor of all models instead.

Ryex replied:
I'm not actually sure what you mean here?

flowln replied:
I think he means that, instead of hardcoding the relative path to the instance dir path, as you have done you, you can instead make `ResourceFolderModel`s constructor receive a `QDir` that's the instance directory, and store it in the class.

You can also, instead of getting a `QDir`, just get a `BaseInstance` reference itself. This may future-proof some stuff :p

Ryex replied:
Ok, that makes sense. the modals now take a shared pointer to a BaseInstance.
This function uses `m_instance->instanceRoot()`

Scrumplex commented on launcher/minecraft/mod/ResourceFolderModel.cpp:
```diff
@@ -415,7 +419,25 @@ QVariant ResourceFolderModel::data(const QModelIndex& index, int role) const
+            if (column == NAME_COLUMN) {
+                if (at(row).isSymLinkUnder(instDirPath())) {
+                    return m_resources[row]->internal_id() +
+                        tr("\nWarning: This resource is symbolically linked from elsewhere. Editing it will also change the original") +
+                        tr("\nCanonical Path: %1").arg(at(row).fileinfo().canonicalFilePath());;
```
wrap in single tr

Scrumplex commented on launcher/minecraft/mod/ResourceFolderModel.cpp:
```diff
@@ -415,7 +419,25 @@ QVariant ResourceFolderModel::data(const QModelIndex& index, int role) const
+                        tr("\nCanonical Path: %1").arg(at(row).fileinfo().canonicalFilePath());;
+                }
+                if (at(row).isMoreThanOneHardLink()) {
+                    return m_resources[row]->internal_id() +
+                        tr("\nWarning: This resource is hard linked elsewhere. Editing it will also change the original");
```
wrap in single tr

Scrumplex commented on launcher/minecraft/mod/ResourcePackFolderModel.cpp:
```diff
@@ -78,12 +82,28 @@ QVariant ResourcePackFolderModel::data(const QModelIndex& index, int role) const
+            if (column == NAME_COLUMN) {
+                if (at(row)->isSymLinkUnder(instDirPath())) {
+                    return m_resources[row]->internal_id() +
+                        tr("\nWarning: This resource is symbolically linked from elsewhere. Editing it will also change the original") +
+                        tr("\nCanonical Path: %1").arg(at(row)->fileinfo().canonicalFilePath());;
```
wrap in single tr

Scrumplex reviewed (commented):
initial review

Ryex commented:
just a heads up, I'm going to be out of a place to live here in a few days. I'll try to finish these changes on my laptop if/when I get a chance but if you don't hear from me for a few weeks feel free to take it over.

Ryex committed:
cleanup: code review sugestions

clean up translation strings

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
workflow: add filelink.exe to SignTool call

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
format: clang-format to fix windows fallout

it looked fine over in vscod on windows but
as soon as I opened it on linux via Helix
the chaos was clear

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex commented:
ok, clearly  they changed something. you can only request review from one person at a time now.

txtsd commented:
> ok, clearly they changed something. you can only request review from one person at a time now.

You should be able to select multiple users. There's no need to uncheck any.

Ryex commented:
> You should be able to select multiple users. There's no need to uncheck any.

I've selected multiple before, but when I select a new reviewer now it automatically unselects the others.

DioEgizio commented:
i think the width should be bigger, on the CoW part it feels like the two text on the bottom are too close to each other
<img width="534" alt="screenshot" src="https://user-images.githubusercontent.com/83089242/224283089-e771f20f-55b4-41a1-b8f0-e4856867d68e.png">

Ryex committed:
fix: adjust geometry and add missing tooltip

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

flowln commented on launcher/CMakeLists.txt:
```diff
@@ -1106,6 +1118,44 @@ install(TARGETS ${Launcher_Name}
 
+if(WIN32)
+    add_library(filelink_logic STATIC ${LINKEXE_SOURCES})
+    target_include_directories(filelink_logic PUBLIC ${CMAKE_CURRENT_SOURCE_DIR})
+    target_link_libraries(filelink_logic
+        systeminfo
+        BuildConfig
+        Qt${QT_VERSION_MAJOR}::Widgets
+        ghcFilesystem::ghc_filesystem
+    )
+    target_link_libraries(filelink_logic
+        Qt${QT_VERSION_MAJOR}::Core
+        Qt${QT_VERSION_MAJOR}::Xml
+        Qt${QT_VERSION_MAJOR}::Network
+        Qt${QT_VERSION_MAJOR}::Concurrent
+        ${Launcher_QT_LIBS}
+    )
```
does it really need to be linked to all those libs? and why is Qt Widgets separate from the other Qt friends?

Ryex replied:
I just copied the link lists from the main launcher. I'm not entirely sure what's necessary to be honest...

Ryex replied:
I'm testing what actualy needs to be here now.

Ryex replied:
turns out, yes. or nearly. only `Qt#::Concurrent` was unneeded

flowln commented on launcher/FileSystem.cpp:
```diff
@@ -215,6 +302,271 @@ bool copy::operator()(const QString& offset, bool dryRun)
+}
+
+/**
+ * @brief make a list off all the  links ot make
```
```suggestion
 * @brief Make a list of all the links to make
```

flowln replied:
i was making these comments since before your newer changes, so feel free to just resolve these outdated ones if they aren't relevant anymore ig

flowln commented on launcher/FileSystem.cpp:
```diff
@@ -215,6 +302,271 @@ bool copy::operator()(const QString& offset, bool dryRun)
+            link_file(src, "");
+        } else {
+            if (m_debug)
+                qDebug() << "linking recursivly:" << src << "to" << dst << "max_depth:" << m_max_depth;
```
```suggestion
                qDebug() << "linking recursively:" << src << "to" << dst << ", max_depth:" << m_max_depth;
```

flowln commented on launcher/FileSystem.cpp:
```diff
@@ -215,6 +302,271 @@ bool copy::operator()(const QString& offset, bool dryRun)
+bool create_link::make_links()
+{
+    for (auto link : m_links_to_make) {
+        QString src_path = link.src;
+        QString dst_path = link.dst;
```
seems like you're using `StringUtils::toStdString` a lot down there. might as well create the converted strings up here and use them down there.

flowln commented on launcher/FileSystem.cpp:
```diff
@@ -279,11 +631,60 @@ QString PathCombine(const QString& path1, const QString& path2, const QString& p
     return QFileInfo(path).absolutePath();
 }
 
+int PathDepth(const QString& path)
```
you could perhaps use [`QDir::cleanPath`](https://doc.qt.io/qt-5/qdir.html#cleanPath) to avoid creating a `QFileInfo` object, and use [`QString::count`](https://doc.qt.io/qt-5/qstring.html#count-2) instead of `split` to get the number of parts, removing the need for the version check?

also, the CamelCase (or pascal case, whatever people call it :p) naming is weird, makes it harder to distinguish between ours and Windows API functions imo, i would like for new functions to use camelCase instead, like most of the rest of the codebase.

Ryex replied:
It's a good thought but without splitting it into parts there isn't a clean way to detect the difference between `".."` and `"."` as a path part. the (now renamed) `pathDepth` function will not always be given a absolute path so not all dot components will be removed. by `QDir::cleanPath`

flowln replied:
The relative path issue could be solved by doing something like this before calling `QDir::cleanPath` (would need to make `path` a copy), if i understand it correctly:
```cpp
if (QDir::isRelativePath(path))
    path = PathCombine(QDir::currentPath(), path);
```

Ryex replied:
It would resolve the path but it would change the behavior of the function.
the path depth I need it to return is the depth from the top of the path given.

flowln commented on launcher/FileSystem.h:
```diff
@@ -186,4 +340,231 @@ bool overrideFolder(QString overwritten_path, QString override_path);
+    {QString("FUSEBLK"), FilesystemType::FUSEBLK},
+    {QString("F2FS"), FilesystemType::F2FS},
+    {QString("UNKNOWN"), FilesystemType::UNKNOWN}
+};
```
why?

the number of FSes seem low enough to not completely justify needing `O(log n)` lookup time instead of having it be `O(n)` with a list of pairs `FilesystemType, QString`, and avoid the data duplication here.

also, these strings should probably be wrapped in the [`QStringLiteral macro`](https://doc.qt.io/qt-5/qstring.html#QStringLiteral)

Ryex replied:
mostly because I was too lazy to write a lookup loop :P.
I'll change it though.

Ryex replied:
Actually... I'm not sure I *should* go with a list of pairs.

this isn't a one to one mapping.

the OS can report many different strings for some of the filesystems (EXT2, EXT3, EXT4 all share the same magic for example)

the Type to string map is one to one but the string to type map is many to one.

and honestly, the isn't used in a function called in a loop. a single hash look up is kinda of fine?

I'll still do the `QStringLiteral` change.

flowln replied:
> this isn't a one to one mapping.

technically, the pair list approach could be a many-to-one mapping if you do a ordered linear search: many filesystem type strings can map to the same type, and a type can map to the first string mapping to it. It could also be achieved with a single map, since it maintains order, but i think a list better exposes the intention of mapping both string -> type and type -> string.

I don't really mind the performance of either approach, i just dislike having two different data structures that need to be managed and maintained separately even though they are intimately related.

Ryex replied:
Getting the many to one mapping from a list isn't the issue really, it's the inversion of the mapping needing a different search order for lookup that gets me.
Id love to deduplicate the structure frankly.

Ryex replied:
ok I think I've found a method I'm satisifed with. `QMap<Enum, QStringList>` where the first elm of the list is the prefered name.

no more duplication.

DioEgizio commented:
Conflicting with develop

Ryex commented:
re-based because clean git history is cool.

Ryex committed:
spelling and formatting

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

flowln commented on launcher/FileSystem.cpp:
```diff
@@ -215,6 +302,271 @@ bool copy::operator()(const QString& offset, bool dryRun)
+        QDataStream out(&block, QIODevice::WriteOnly);
+        out.setVersion(QDataStream::Qt_5_0);  // choose correct version better?
+
+        qint32 blocksize = quint32(sizeof(quint32));
```
why do you need the blocksize on either end of the communication?

Ryex replied:
This is raw socket communication. it is necessary to specify how much data we are expecting to read back out as one if not the first thing in the packet. it's a necessary check to ensure that all the bytes that need to be consumed are present and accoutered for.

flowln replied:
It is using Qt's abstraction though, wouldn't any problem of this sort trigger the error stuff in `QLocalSocket`?

Ryex replied:
nope. No such guarantees are provided. This practice is pulled directly form Qt's examples on how to use QLocalSocket.

flowln replied:
i see... sadge

flowln commented on launcher/FileSystem.cpp:
```diff
@@ -215,6 +302,271 @@ bool copy::operator()(const QString& offset, bool dryRun)
+        // construct block of data to send
+        QByteArray block;
+        QDataStream out(&block, QIODevice::WriteOnly);
+        out.setVersion(QDataStream::Qt_5_0);  // choose correct version better?
```
why do you need to set a version? I think it should use the default one, since it's the most likely to work better, and since both ends of the communication have the same Qt version, they will choose the same default every time.

Ryex replied:
The types are being encoded as raw bytes and set over a socket in order for the under end to know how to reconstruct the types on the other end both sides have to agree on a format for the bytes to be in. each version of QT has new or changed types and so needs a new protocol version. by setting a version here Qt knows how to encode and decode the byte data.

flowln replied:
Yes, but both sides of the communication are using `QDataStream` in the very same Qt version. It may be a raw binary encoding, but `QDataStream` will always (de)serialize that data in the same way on both sides, so it shouldn't matter which exact version we're using, provided we'll use the same one as the default anyway.

Ryex replied:
choosing a protocol version that is stable and leaving it seems like a good solution. this way it won't unexpectedly break if we have an update and for some reason one binary isn't updated correctly.

flowln replied:
> this way it won't unexpectedly break if we have an update and for some reason one binary isn't updated correctly

how would that even happen... I'm more worried about fixing the version and that specific protocol version being accidentally broken in a new Qt version without anyone noticing it, since there's so many of them. I would assume it would be harder for that to happen in the default version, since that's the most likely to be tested.

Ryex replied:
I suppose you have a point. I was more worried that if there was an update and the main launcher updated by the link binary didn't it would break. but that is a fixable problem I guess.

I went ahead and removed the explicit version.

flowln commented on launcher/FileSystem.cpp:
```diff
@@ -215,6 +302,271 @@ bool copy::operator()(const QString& offset, bool dryRun)
+            qDebug() << "Reading path results from client";
+            qDebug() << "bytes available" << clientConnection->bytesAvailable();
+
+            // Relies on the fact that QDataStream serializes a quint32 into
+            // sizeof(quint32) bytes
+            if (clientConnection->bytesAvailable() < (int)sizeof(quint32))
+                return;
```
why do you need that?

Ryex replied:
it is theoretically possible for a packet of data from one client to not arrive complete and we need to wait for the rest of the bytes before attempting to decode the data from the stream.
In practice at the level of local sockets something would need to go horridly wrong or someone would need to be trying something jank like proxying the socket. Still it's a good idea to have this here just in case.

flowln commented on launcher/settings/INIFile.cpp:
```diff
@@ -183,3 +183,21 @@ void INIFile::set(QString key, QVariant val)
     this->operator[](key) = val;
 }
+
+void INIFile::setList(QString key, QVariantList val)
+{
+    QString stringList = QJsonDocument(QVariant(val).toJsonArray()).toJson(QJsonDocument::Compact);
+
+    this->operator[](key) = stringList;
+}
+
+QVariantList INIFile::getList(QString key, QVariantList def) const
+{
+    if (this->contains(key)) {
+        auto src = this->operator[](key);
+
+        return QJsonDocument::fromJson(src.toByteArray()).toVariant().toList();
+    }
+
+    return def;
+}
```
Why are those needed when `QVariant` accepts `QVariantList` too?

Ryex replied:
I played around with a LOT of ways to pull a list out of an INI value. This is one of the only ways the compiler would accept. everything else complained about `QVariant` not understanding the underlying type. Technically now that I understand how `QMetaType`'s work I could probably resolve this but frankly this seems clean in comparason.

flowln replied:
I think having separate functions for `QVariant` and `QVariantList` is confusing, especially considering `QVariantList` is technically accepted in `QVariant`.

Ryex replied:
the problem is that there is no real conversion between a `QStringList` and a `QVariant` or even a `QVariantList`
also the underlying `toString` that the INI file class used to save things doesn't understand how to save lists. The one time I got the type conversions to work I ended up with blank settings.

The JSON conversion is necessary and this necessitates the split between `get` and `getList`

flowln replied:
After some tests, it seems that the only thing preventing `get` and `set` from working with those lists is the `INIFile` implementation thing. I think I said it before, but that thing is trash and I will stand by it until the dawn of time.
I have made a simple modification on it so that it's actually just a wrapper around a `QSettings` object, and it seems to work fine. I think it's a better alternative than adding a separate function and doing the JSON thing:

<details>
<summary>Le patch (note that I didn't remove the `getList` method, and neither changed the usages of it in `BaseInstance` and the test. Also, i didn't look at the `escape` and `unenscape` methods in INIFile, but maybe they can also be :hocho: )</summary>

```patch
diff --git a/launcher/minecraft/mod/tasks/LocalModParseTask.cpp b/launcher/minecraft/mod/tasks/LocalModParseTask.cpp
index da27a505..5342d693 100644
--- a/launcher/minecraft/mod/tasks/LocalModParseTask.cpp
+++ b/launcher/minecraft/mod/tasks/LocalModParseTask.cpp
@@ -242,7 +242,7 @@ ModDetails ReadQuiltModInfo(QByteArray contents)
     return details;
 }
 
-ModDetails ReadForgeInfo(QByteArray contents)
+ModDetails ReadForgeInfo(QString fileName)
 {
     ModDetails details;
     // Read the data
@@ -250,7 +250,7 @@ ModDetails ReadForgeInfo(QByteArray contents)
     details.mod_id = "Forge";
     details.homeurl = "http://www.minecraftforge.net/forum/";
     INIFile ini;
-    if (!ini.loadFile(contents))
+    if (!ini.loadFile(fileName))
         return details;
 
     QString major = ini.get("forge.major.number", "0").toString();
@@ -422,7 +422,7 @@ bool processZIP(Mod& mod, ProcessingLevel level)
             return false;
         }
 
-        details = ReadForgeInfo(file.readAll());
+        details = ReadForgeInfo(file.getFileName());
         file.close();
         zip.close();

diff --git a/launcher/settings/INIFile.cpp b/launcher/settings/INIFile.cpp
index e48e6f47..f7588fb8 100644
--- a/launcher/settings/INIFile.cpp
+++ b/launcher/settings/INIFile.cpp
@@ -42,6 +42,8 @@
 #include <QSaveFile>
 #include <QDebug>
 
+#include <QSettings>
+
 INIFile::INIFile()
 {
 }
@@ -99,75 +101,43 @@ QString INIFile::escape(QString orig)
 
 bool INIFile::saveFile(QString fileName)
 {
-    QByteArray outArray;
+    QSettings _settings_obj{ fileName, QSettings::Format::IniFormat };
+    _settings_obj.setFallbacksEnabled(false);
+
     for (Iterator iter = begin(); iter != end(); iter++)
-    {
-        QString value = iter.value().toString();
-        value = escape(value);
-        outArray.append(iter.key().toUtf8());
-        outArray.append('=');
-        outArray.append(value.toUtf8());
-        outArray.append('\n');
-    }
+        _settings_obj.setValue(iter.key(), iter.value());
+
+    _settings_obj.sync();
+
+    if (auto status = _settings_obj.status(); status != QSettings::Status::NoError) {
+        // Shouldn't be possible!
+        Q_ASSERT(status != QSettings::Status::FormatError);
+
+        if (status == QSettings::Status::AccessError)
+            qCritical() << "An access error occurred (e.g. trying to write to a read-only file).";
 
-    try
-    {
-        FS::write(fileName, outArray);
-    }
-    catch (const Exception &e)
-    {
-        qCritical() << e.what();
         return false;
     }
 
     return true;
 }
 
-
 bool INIFile::loadFile(QString fileName)
 {
-    QFile file(fileName);
-    if (!file.open(QIODevice::ReadOnly))
+    QSettings _settings_obj{ fileName, QSettings::Format::IniFormat };
+    _settings_obj.setFallbacksEnabled(false);
+
+    if (auto status = _settings_obj.status(); status != QSettings::Status::NoError) {
+        if (status == QSettings::Status::AccessError)
+            qCritical() << "An access error occurred (e.g. trying to write to a read-only file).";
+        if (status == QSettings::Status::FormatError)
+            qCritical() << "A format error occurred (e.g. loading a malformed INI file).";
         return false;
-    bool success = loadFile(file.readAll());
-    file.close();
-    return success;
-}
-
-bool INIFile::loadFile(QByteArray file)
-{
-    QTextStream in(file);
-#if QT_VERSION <= QT_VERSION_CHECK(6, 0, 0)
-    in.setCodec("UTF-8");
-#endif
-
-    QStringList lines = in.readAll().split('\n');
-    for (int i = 0; i < lines.count(); i++)
-    {
-        QString &lineRaw = lines[i];
-        // Ignore comments.
-        int commentIndex = 0;
-        QString line = lineRaw;
-        // Search for comments until no more escaped # are available
-        while((commentIndex = line.indexOf('#', commentIndex + 1)) != -1) {
-            if(commentIndex > 0 && line.at(commentIndex - 1) == '\\') {
-                continue;
-            }
-            line = line.left(lineRaw.indexOf('#')).trimmed();
-        }
-
-        int eqPos = line.indexOf('=');
-        if (eqPos == -1)
-            continue;
-        QString key = line.left(eqPos).trimmed();
-        QString valueStr = line.right(line.length() - eqPos - 1).trimmed();
-
-        valueStr = unescape(valueStr);
-
-        QVariant value(valueStr);
-        this->operator[](key) = value;
     }
 
+    for (auto&& key : _settings_obj.allKeys())
+        insert(key, _settings_obj.value(key));
+
     return true;
 }
 
diff --git a/launcher/settings/INIFile.h b/launcher/settings/INIFile.h
index 86bf0898..3745acd2 100644
--- a/launcher/settings/INIFile.h
+++ b/launcher/settings/INIFile.h
@@ -28,7 +28,6 @@ class INIFile : public QMap<QString, QVariant>
 public:
     explicit INIFile();
 
-    bool loadFile(QByteArray file);
     bool loadFile(QString fileName);
     bool saveFile(QString fileName);
```

</details>

Ryex replied:
I'm happy to apply this patch but it still leaves me needing a template function to set and get `QList<T>` where `T` is not `QString`

there is no automatic conversion to `QVariant` for `QList<int>` for example

```cpp
template <typename T> QList<T>  INIFile::getList(QString key, QList<T> def) const
{   
    if (this->contains(key)) {
        QVariant src = this->operator[](key);
        QVariantList variantList = src.toList();

        QList<T> list_t;
        list_t.reserve(variantList.size());
        for (const QVariant& v : variantList)
        {
            list_t.append(v.value<T>());
        }
        return list_t;
    }

    return def;   
}

template <typename T> void INIFile::setList(QString key, QList<T> val)
{
    QVariantList variantList;
    variantList.reserve(val.size());
    for (const T& v : val)
    {
        variantList.append(v);
    }

    this->set(key, variantList);
}
```

Ryex replied:
Patch applied, list functions moved to `QVariantHelpers`

flowln commented on launcher/settings/SettingsObject.cpp:
```diff
@@ -121,6 +121,19 @@ bool SettingsObject::contains(const QString &id)
+{   
+    QVariant value = this->get(id);
+    return QJsonDocument::fromJson(value.toByteArray()).toVariant().toList();
+}
```
same here, why not send the `QVariantList` directly?

flowln commented on launcher/ui/MainWindow.cpp:
```diff
@@ -1337,6 +1337,20 @@ void MainWindow::on_actionDeleteInstance_triggered()
+    if (!linkedInstances.empty()) {
+        response = CustomMessageBox::selectable(
+                this, tr("There are linked instances"),
+                tr("The folowing Instance(s) might reference files in this instance:\n\n"
```
```suggestion
                tr("The folowing instance(s) might reference files in this instance:\n\n"
```

flowln commented on launcher/ui/MainWindow.cpp:
```diff
@@ -1337,6 +1337,20 @@ void MainWindow::on_actionDeleteInstance_triggered()
+                tr("The folowing Instance(s) might reference files in this instance:\n\n"
+                "%1\n\n"
+                "Deleting it could break the other instance(s), \n\n"
+                "Are you sure?").arg(linkedInstances.join("\n")), 
```
```suggestion
                "Do you wish to proceed?").arg(linkedInstances.join("\n")), 
```

flowln reviewed (commented):
> i may be able to finish it off in the weekend

i lied

sorry, this PR gives me a lot of brain fatigue (your work is great, but os-specific stuff gives me nightmares) :dizzy_face:

Ryex committed:
fix: Apply suggestions from code review (string changes)

Co-authored-by: flow <flowlnlnln@gmail.com>
Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix: typos, CamelCase  to camelCase the new names

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex commented:
> > i may be able to finish it off in the weekend
> 
> i lied
> 
> sorry, this PR gives me a lot of brain fatigue (your work is great, but os-specific stuff gives me nightmares) dizzy_face

no worries. don't feel too overwhelmed by it though. when it comes down to it it's pretty simple stuff despite appearances. I documented the parts that need external research to really understand.

Ryex committed:
refactor: remove data duplication in statis FS Names

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix: remove fixed datastream version

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix: add back QT::Widgets link

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix: filelink needs network for local socket

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix: sysinfo libs needed too

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Continous commented:
> @getchoo @TayouVR I've added a somewhat naive link tracking system. each instance when it is made via copy saves a `linkedInstancesList` setting with is a json list containing the ID of the original instance. When an instance is asked to be deleted there is now an additional second message box that lists the instances that say they are linked to the deletion candidate, warns that these instances may break, and asks for confirmation.
> 
> @DioEgizio I made sure the `_filelink.exe` was included in the setup exe, I know you mentioned that somewhere.

Is there any possibility we could have an additional option to allow a deletion of the original instance to prompt the user to have the original files migrated to the linked instance? If it's not too difficult, we could even allow linked instances to become "real" instances on command from the user in such case they may want or need it. A good example I can think of is if someone plans on migrating from one version to the next, and has finally decided to remove the previous instance.

Ryex commented:
> > @getchoo @TayouVR I've added a somewhat naive link tracking system. each instance when it is made via copy saves a `linkedInstancesList` setting with is a json list containing the ID of the original instance. When an instance is asked to be deleted there is now an additional second message box that lists the instances that say they are linked to the deletion candidate, warns that these instances may break, and asks for confirmation.
> > @DioEgizio I made sure the `_filelink.exe` was included in the setup exe, I know you mentioned that somewhere.
> 
> Is there any possibility we could have an additional option to allow a deletion of the original instance to prompt the user to have the original files migrated to the linked instance? If it's not too difficult, we could even allow linked instances to become "real" instances on command from the user in such case they may want or need it. A good example I can think of is if someone plans on migrating from one version to the next, and has finally decided to remove the previous instance.

perhaps in another PR. that's a fair bit of work and this one is big enough already.

flowln commented on launcher/FileSystem.cpp:
```diff
@@ -215,6 +302,272 @@ bool copy::operator()(const QString& offset, bool dryRun)
+/**
+ * @brief Make a list of all the links to make
+ * @param offset subdirectory form src to link to dest
+ * @return if there was an error during the attempt to link
```
```suggestion
 * @param offset subdirectory of src to link to dest
```

flowln commented on launcher/ui/MainWindow.cpp:
```diff
@@ -1337,6 +1337,20 @@ void MainWindow::on_actionDeleteInstance_triggered()
+    if (!linkedInstances.empty()) {
+        response = CustomMessageBox::selectable(
+                this, tr("There are linked instances"),
+                tr("The folowing instance(s) might reference files in this instance:\n\n"
+                "%1\n\n"
+                "Deleting it could break the other instance(s), \n\n"
+                "Do you wish to proceed?").arg(linkedInstances.join("\n")), 
```
```suggestion
                tr("The following instance(s) might reference files in this instance:\n\n"
                "%1\n\n"
                "Deleting it could break the other instance(s), \n\n"
                "Do you wish to proceed?", nullptr, linkedInstances.count()).arg(linkedInstances.join("\n")), 
```

Ok, I actually don't know for sure this is how it's done, but my interpretation of https://doc.qt.io/qt-6/i18n-source-translation.html#handle-plural-forms says this is how you do it :cold_sweat:

flowln commented on launcher/ui/dialogs/CopyInstanceDialog.cpp:
```diff
@@ -85,6 +87,35 @@ CopyInstanceDialog::CopyInstanceDialog(InstancePtr original, QWidget *parent)
+#if defined(Q_OS_WIN)
+    ui->symbolicLinksCheckbox->setIcon(style()->standardIcon(QStyle::SP_VistaShield));
+    ui->symbolicLinksCheckbox->setToolTip(tr("Use symbolic links instead of copying files.") +
+                                          tr("\nOn windows symbolic links may require admin permission to create."));
```
```suggestion
                                          "\n" + tr("On Windows, symbolic links may require admin permission to create."));
```

flowln reviewed (commented):
i swear we will finish this soon :pleading_face:

Ryex committed:
refactor: Move ini to use QSettings && drop get/setList functions

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
refactor: pass instance ptr to resource models. use it to find instance root.

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

flowln reviewed (approved):
good job :slightly_smiling_face: /gen

DioEgizio commented:
2 approvals 👀

Scrumplex commented:
Let's merge, I guess.

I probably don't have much to add after @flowln's thorough analysis :D

And if the feature has issues, we can hopefully catch them before an eventual release.

Un1q32 commented:
1.20 prerelease 7 gives a warning when opening worlds that use symlinks, and requires allowed symlinks to be added to a text file to silence the warning.  Will this cause problems?

DioEgizio commented:
Make an issue for it please

Un1q32 commented:
> Make an issue for it please

I haven't actually tested anything so I don't know if its a problem, I just randomly remembered it so figured I would mention it.

RNanB commented:
> 1.20 prerelease 7 gives a warning when opening worlds that use symlinks, and requires allowed symlinks to be added to a text file to silence the warning. Will this cause problems?

No, this change only affects linking of individual worlds. Linking the saves folder itself will still work the same

Ryex commented:
> > 1.20 prerelease 7 gives a warning when opening worlds that use symlinks, and requires allowed symlinks to be added to a text file to silence the warning. Will this cause problems?
> 
> No, this change only affects linking of individual worlds. Linking the saves folder itself will still work the same

This could still have an effect if the user tries to recursively link during a copy. That *will* need the file created

ToppyTopology commented:
hello, is there a way to deselect the config folder from instance copy options? or is this a limitation since it looks like its apart of copy mods option

Trial97 commented:
Just deselect it inside the export dialog:
![image](https://github.com/PrismLauncher/PrismLauncher/assets/42720688/93847c40-8994-440e-942d-ed9b3a118119)

Also please refrain from asking questions on a closed issue/pull request more so if that are not related.
If you have more questions join discord: https://prismlauncher.org/discord
